executor/slow_query.go

```diff
@@ -118,16 +116,53 @@ func (e *slowQueryRetriever) close() error {
 	return nil
 }
 
-func (e *slowQueryRetriever) dataForSlowLog(ctx sessionctx.Context) ([][]types.Datum, error) {
-	reader := bufio.NewReader(e.files[e.fileIdx].file)
-	rows, err := e.parseSlowLog(ctx, reader, 1024)
+type parsedSlowLog struct {
+	rows [][]types.Datum
+	err  error
+}
+
+func (e *slowQueryRetriever) parseDataForSlowLog(ctx context.Context, sctx sessionctx.Context) {
+	if len(e.files) == 0 {
+		close(e.parsedSlowLogCh)
+		return
+	}
```

Comment:
It this judgment redundant?

Reply:
If we want to initialize the reader once, one way is to put it outside of the loop. In this way, we need to make sure we have at least one file to read, otherwise, we cannot get `e.files[0]`. So, we need this condition statement.

The other way is to put the reader's first initialization inside the loop but guard it with a condition statement like this.
```go
    var reader *bufio.Reader
	for e.fileIdx < len(e.files) {
        if reader == nil {
                reader = bufio.NewReader(e.files[e.fileIdx].file)
        }
 		rows, err := e.parseSlowLog(sctx, reader, 1024)
 		select {
 		case <-ctx.Done():
 			break
 		case e.parsedSlowLogCh <- parsedSlowLog{rows, err}:
 		}
 	}
```
I thinks the first way is clearer and more effiecient when the number of log files is large.

Reply:
Make sence. Thanks for the explanation.